pkg/providers/v1/aws_test.go

```diff
@@ -3741,3 +3742,86 @@ func TestGetZoneByProviderIDForFargate(t *testing.T) {
 	assert.Nil(t, err)
 	assert.Equal(t, "us-west-2c", zoneDetails.FailureDomain)
 }
+
+type MockedEC2API struct {
+	ec2iface.EC2API
+	mock.Mock
+}
+
+func (m *MockedEC2API) DescribeInstances(input *ec2.DescribeInstancesInput) (*ec2.DescribeInstancesOutput, error) {
+	args := m.Called(input)
+	return args.Get(0).(*ec2.DescribeInstancesOutput), args.Error(1)
+}
+
+func newMockedEC2API() *MockedEC2API {
+	return &MockedEC2API{}
+}
+
+func TestDescribeInstances(t *testing.T) {
+	tests := []struct {
+		name    string
+		input   *ec2.DescribeInstancesInput
+		expect  func(interface{})
```

Comment:
why is it interface{}? what is expect supposed to do? what would a failure look like? if in every test case I tell DescribeInstances what to take as input and what to return, what is being tested?

Reply:
I think I'm missing an `AssertExpectations` call in the test (I was a little confused as to whether I needed it while writing the test, because the test does fail if DescribeInstances is not called as expected).   Let me see if I can determine if its needed and clear up the logic.

Reply:
Documentation for On: https://pkg.go.dev/github.com/stretchr/testify/mock#Call.On
so I am telling it what to return when given so and so input. And I think ultimately I have to assert that this call happened at all by doing https://pkg.go.dev/github.com/stretchr/testify/mock#Mock.AssertExpectations 

(new to this package, slightly more familiar with gomock)

Reply:
we dont need Assert since Called basically does it for us . https://pkg.go.dev/github.com/stretchr/testify/mock?utm_source=godoc#Mock.Called

Reply:
Right, that's the reason that I didn't have an explicit call to AssertExpectations -- because I was seeing tests fail if I called the expected functions with unexpected arguments.  That being said, I think we do want an explicit call to `AssertExpectations` because of [the missing case](https://github.com/stretchr/testify/blob/e209ca88af1919e710960205e887e57db33c9be0/mock/mock.go#L529-L531). If there is an assert, and a missing call, it will only be caught by the explicit `AssertExpectations`, whereas `Called` will catch calls with improper arguments.